File: nim/ads/queue.nim
const ARRAY_SIZE = 5

type Queue = object
  items: array[ARRAY_SIZE, int]
  front: int
  rear: int

proc newQueue(): Queue = 
  Queue(front: -1, rear: 0)

proc isEmpty(queue: Queue): bool =
  queue.front == (queue.rear - 1)

proc isFull(queue: Queue): bool =
  queue.rear == ARRAY_SIZE

proc enqueue(queue: var Queue, value: int) = 
  if (queue.isFull()):
    echo "Queue is full"
    return

  queue.items[queue.rear] = value
  queue.rear += 1
  echo "Enqueue: " & $value

proc dequeue(queue: var Queue) =
  if (queue.isEmpty()):
    echo "Queue is empty"
    return

  queue.front += 1

proc peek(queue: var Queue) =
  if (queue.isEmpty()):
    echo "Queue is empty"

  queue.front += 1
  echo "Picked: " & $queue.items[queue.front]

proc print(queue: Queue) =
  if (queue.isEmpty()):
    echo "Queue is empty"
    return

  echo "Current Queue: "
  for i in queue.front+1 ..< queue.rear:
    echo queue.items[i]
  echo()

proc main() =
  var q = newQueue()
  q.enqueue(10)
  q.enqueue(20)
  q.enqueue(30)
  q.enqueue(40)
  q.enqueue(50)
  q.enqueue(60)
  # q.print()

  echo()
  q.peek()
  q.peek()
  q.peek()
  # q.dequeue()

  echo()
  q.print()

when isMainModule:
  main()

File: clang/ads/queue.c
#include <stdbool.h>
#include <stdio.h>
#include <stdbool.h>

#define MAX_SIZE 100

typedef struct {
  int items[MAX_SIZE];
  int front;
  int rear;
} Queue;

bool  isfull(Queue* queue);
bool  isempty(Queue* queue);
int   peek(Queue* queue);
void  init(Queue* queue);
void  print(Queue* queue);
void  dequeue(Queue* queue);
void  enqueue(Queue* queue, int value);

void init(Queue* queue){
  queue->front = -1;
  queue->rear = 0;
}

bool isempty(Queue* queue){
  return queue->front == (queue->rear - 1);
}

bool isfull(Queue* queue){
  return queue->rear >= MAX_SIZE;
}

void enqueue(Queue* queue, int value){
  if (isfull(queue)){
    printf("Queue is full\n");
    return;
  }

  queue->items[queue->rear] = value;
  queue->rear++;
}

void dequeue(Queue* queue){
  if (isempty(queue)){
    printf("Queue is empty\n");
    return;
  }
  queue->front++;
}

int peek(Queue* queue){
  if (isempty(queue)){
    printf("Queue is empty\n");
    return -1;
  }
  return queue->items[queue->front + 1];
}

void print(Queue* queue){
  if (isempty(queue)){
    printf("Queue is empty\n");
    return;
  }

  printf("Current Queue: ");
  for (int i = queue->front + 1; i < queue->rear; i++){
    printf("%d ", queue->items[i]);
  }

  printf("\n");
}
